src/banded/BandedMatrix.jl

```diff
@@ -43,22 +43,24 @@ const BandedSubBandedMatrix{T} =
 
 returns an uninitialized `n`×`m` banded matrix of type `T` with bandwidths `(l,u)`.
 """
-BandedMatrix{T}(::UndefInitializer, n::Integer, m::Integer, a::Integer, b::Integer) where {T<:BlasFloat} =
-    _BandedMatrix(Matrix{T}(undef,max(0,b+a+1),m), n, a, b)
-BandedMatrix{T}(::UndefInitializer, n::Integer, m::Integer, a::Integer, b::Integer)  where {T<:Number} =
+BandedMatrix{T, C}(::UndefInitializer, n::Integer, m::Integer, a::Integer, b::Integer) where {T<:BlasFloat, C<:AbstractMatrix{T}} =
+    _BandedMatrix(C(undef,max(0,b+a+1),m), n, a, b)
+BandedMatrix{T, C}(::UndefInitializer, n::Integer, m::Integer, a::Integer, b::Integer) where {T, C<:AbstractMatrix{T}} =
+    _BandedMatrix(C(undef,max(0,b+a+1),m),n,a,b)
+BandedMatrix{T, C}(::UndefInitializer, n::Integer, m::Integer, a::Integer, b::Integer)  where {T<:Number, C<:AbstractMatrix{T}} =
     _BandedMatrix(zeros(T,max(0,b+a+1),m),n,a,b)
 BandedMatrix{T}(::UndefInitializer, n::Integer, m::Integer, a::Integer, b::Integer)  where {T} =
-    _BandedMatrix(Matrix{T}(undef,max(0,b+a+1),m),n,a,b)
+    BandedMatrix{T, Matrix{T}}(undef,n,m,a,b)
 BandedMatrix{T}(::UndefInitializer, nm::NTuple{2,Integer}, ab::NTuple{2,Integer}) where T =
     BandedMatrix{T}(undef, nm..., ab...)
 BandedMatrix{T}(::UndefInitializer, n::Integer, ::Colon, a::Integer, b::Integer)  where {T} =
     BandedMatrix{T}(undef,n,n+b,a,b)
 
 
 BandedMatrix{V}(M::BandedMatrix) where {V} =
-        _BandedMatrix(AbstractMatrix{V}(M.data), M.m, M.l, M.u)
+    _BandedMatrix(copy!(similar(M.data, V), M.data), M.m, M.l, M.u)
```

Comment:
I think this should be left as `AbstractMatrix{V}(M.data)`:  `M.data` may be an immutable type, and so `copy!` is an invalid operation.

Reply:
You're right about not using `copy!`. But I don't think we want to instantiate `BandedMatrix{Float, SomeAbstractType}` . That would yield slow code.

I'll see what I can figure out.

Reply:
mmm, looking at it in more detail, it seems you are right. This is somewhat different from casting to a base class in C++. I'll get rid of that commit.